### PR TITLE
RTPS: msg generation: generalize scripts

### PR DESCRIPTION
**Pull request by TSC21**
This PR generalizes the scripts fro generating the RTPS related headers and source files.
1. Allows passing the templates as arguments, so to allow the scripts to use different templates;
2. Allows passing the pkg name, which is useful when setting field types which depends on other msg types that belong to a certain package - it also allows using the scripts out of the PX4 Firmware repo;
3. Sets the RTPS msg ID's to the templates by loading these from a JSON file - this allows setting other JSON files with other IDs or message naming (which for the ROS side is required, since the msg naming is different).

### CODE REVIEW

**Comment by LorenzMeier** [comment]
Great! What do we need to wrap this up so we can start to use it on a daily basis?

**Comment by TSC21** [comment]
Still cleaning it up, but when you talk about use it on a daily basis you mean the RTPS<->ROS bridge? If that's so, it's a matter of pushing this in and bring some last changes on the build process for the `px4_ros_com` package and should be good to go. In terms of general usage of RTPS, I think it's a matter of updating the the documentation for a more clear view of what's required and the current status, which is in my todo list as well.

File: msg/templates/urtps/RtpsTopics.cpp.template [hunk 1]
```diff
@@ -88,7 +89,7 @@ void RtpsTopics::publish(uint8_t topic_ID, char data_buffer[], size_t len)
     switch (topic_ID)
     {
 @[for topic in send_topics]@
-        case @(message_id(topic)): // @(topic)
+        case @(rtps_message_id(os.path.abspath('uorb_rtps_message_ids.json'), topic)): // @(topic)
```
Comment (bkueng) [inline msg/templates/urtps/RtpsTopics.cpp.template]: This needs to parse a json file for every message. Can we instead pass an already parsed `dict` to the template?

File: msg/tools/generate_microRTPS_bridge.py [hunk 2]
```diff
@@ -57,13 +57,21 @@ def get_absolute_path(arg_parse_dir):
...
 parser.add_argument("-a", "--agent", dest='agent', action="store_true", help="Flag for generate the agent, by default is true if -c is not specified")
 parser.add_argument("-c", "--client", dest='client', action="store_true", help="Flag for generate the client, by default is true if -a is not specified")
+parser.add_argument("-i", "--no-idl", dest='idl', action="store_false", help="Flag for generate idl files for each msg, by default is true if -i is not specified")
 parser.add_argument("-t", "--topic-msg-dir", dest='msgdir', type=str, nargs=1, help="Topics message dir, by default msg/", default="msg")
+parser.add_argument("-b", "--uorb-templates-dir", dest='uorb_templates', type=str, nargs=1, help="uORB templates dir, by default msg_dir/templates/uorb_microcdr", default=default_uorb_templates_dir)
+parser.add_argument("-q", "--urtps-templates-dir", dest='urtps_templates', type=str, nargs=1, help="uRTPS templates dir, by default msg_dir/templates/urtps", default=default_urtps_templates_dir)
+parser.add_argument("-p", "--package", dest='package', type=str, nargs=1, help="Msg package naming, by default px4", default=default_package_name)
```
Comment (bkueng) [inline msg/tools/generate_microRTPS_bridge.py]: You can drop `nargs=1`, then you don't get a list.

File: msg/tools/generate_microRTPS_bridge.py [hunk 3]
```diff
@@ -77,18 +85,26 @@ def get_absolute_path(arg_parse_dir):
...
+if args.package[0] != px_generate_uorb_topic_files.PACKAGE:
+    package = args.package[0]
 else:
-    msg_files_receive = []
+    package = px_generate_uorb_topic_files.PACKAGE
+
 agent = args.agent
 client = args.client
+# If not specified, auto generate the idl filename_send_msgs
+if args.idl is None or args.idl == "":
```
Comment (bkueng) [inline msg/tools/generate_microRTPS_bridge.py]: If you add a `default=False` to `parser.add_argument("-i", "--no-idl"`, you can directly set `idl = args.idl`

File: msg/templates/urtps/RtpsTopics.cpp.template (hunk 1 above)
Comment (TSC21) [reply msg/templates/urtps/RtpsTopics.cpp.template]: Makes sense. Will do that then.

File: msg/tools/generate_microRTPS_bridge.py (hunk 2 above)
Comment (TSC21) [reply msg/tools/generate_microRTPS_bridge.py]: sure. thanks

File: msg/tools/generate_microRTPS_bridge.py (hunk 3 above)
Comment (TSC21) [reply msg/tools/generate_microRTPS_bridge.py]: true. thanks

**Comment by TSC21** [comment]
> We could switch to yaml instead of json, as it is easier to edit and ROS uses it as well.

Sounds reasonable. Will use `yaml` instead.

File: msg/tools/uorb_rtps_message_ids.json [hunk 4]
```diff
@@ -0,0 +1,352 @@
+{
+  "rtps": [
+    {
+      "msg": "actuator_armed",
+      "id": 1
```
Comment (dagar) [inline msg/tools/uorb_rtps_message_ids.json]: Could we do this as a metadata field in the message itself? Alternatively something in the CMakeLists.txt might also make sense. https://github.com/PX4/Firmware/blob/master/msg/CMakeLists.txt#L35-L140
I'm worried about maintaining the list over time.

**Comment by TSC21** [comment]
@dagar seems we are missing `pyyaml` on OSX: http://ci.px4.io:8080/blue/organizations/jenkins/PX4_misc%2FFirmware-compile/detail/PR-10388/4/pipeline/42

**Comment by TSC21** [comment]
@bkueng everything addressed and now compiling. Thanks for the tips. Now we just need https://github.com/PX4/containers/pull/133 in.

File: msg/templates/urtps/RtpsTopics.h.template [hunk 5]
```diff
@@ -9,15 +9,24 @@
...
 import genmsg.msgs
 import gencpp
 from px_generate_uorb_topic_helper import * # this is in Tools/
 from px_generate_uorb_topic_files import MsgScope # this is in Tools/
-from uorb_rtps_message_ids import * # this is in Tools/
 
 topic_names = [single_spec.short_name for single_spec in spec]
 send_topics = [s.short_name for idx, s in enumerate(spec) if scope[idx] == MsgScope.SEND]
 recv_topics = [s.short_name for idx, s in enumerate(spec) if scope[idx] == MsgScope.RECEIVE]
+yaml_file = os.path.abspath('uorb_rtps_message_ids.yaml')
```
[the remark this answers is not in the document]
Comment (TSC21) [reply msg/templates/urtps/RtpsTopics.h.template]: Makes sense. I will investigate further on how to do it.

File: src/modules/micrortps_bridge/micrortps_client/CMakeLists.txt [hunk 6]
```diff
@@ -71,14 +71,23 @@ if (NOT "${config_rtps_send_topics}" STREQUAL "" OR NOT "${config_rtps_receive_t
 		${msg_out_path}/micrortps_client/microRTPS_transport.h
 		)
 
+	# Copy yaml file to the bin dir to be used by empy on the file generation
+	configure_file("${PX4_SOURCE_DIR}/msg/tools/uorb_rtps_message_ids.yaml"
```
[the remark this answers is not in the document]
Comment (TSC21) [reply src/modules/micrortps_bridge/micrortps_client/CMakeLists.txt]: Unless we set it manually on the template file or pass it through `globals` as you suggested, this was the only way I found, as the generator code makes a copy of the template and generates the code on the bin dir, so it's as the python script is run on the same dir, which leads to `os.path.abspath('uorb_rtps_message_ids.yaml')` to look on that same dir. I will look further for a solution here, but wouldn't want to spend much more time on it.

File: msg/tools/px_generate_uorb_topic_helper.py [hunk 7]
```diff
@@ -4,6 +4,9 @@
 # Another positive effect of having the code here, is that this file will get
 # precompiled and thus message generation will be much faster
 
+import os
+import yaml
```
[the remark this answers is not in the document]
Comment (TSC21) [reply msg/tools/px_generate_uorb_topic_helper.py]: Ok sure.

**Comment by TSC21** [comment]
> Looks good, I added a some remarks.

@bkueng I will try to address your remarks but this will also cause some changes on the `px4_ros_com` package I am developing, which would delay it's progress. Can these changes like the `'uorb_rtps_message_ids.yaml'` dir be addressed on another PR, or do you see them as required at this step?

File: msg/tools/generate_microRTPS_bridge.py [hunk 8]
```diff
@@ -77,20 +87,30 @@ def get_absolute_path(arg_parse_dir):
...
-else:
-    msg_files_send = []
 if args.receive:
     msg_files_receive = [get_absolute_path(msg) for msg in args.receive]
+
+if args.package != px_generate_uorb_topic_files.PACKAGE:
+    package = args.package
 else:
-    msg_files_receive = []
+    package = px_generate_uorb_topic_files.PACKAGE
```
[the remark this answers is not in the document]
Comment (TSC21) [reply msg/tools/generate_microRTPS_bridge.py]: done

File: msg/tools/generate_microRTPS_bridge.py [hunk 9]
```diff
@@ -77,20 +87,30 @@ def get_absolute_path(arg_parse_dir):
...
-px_generate_uorb_topic_files.append_to_include_path({msg_folder}, px_generate_uorb_topic_files.INCL_DEFAULT)
+px_generate_uorb_topic_files.append_to_include_path({msg_folder}, px_generate_uorb_topic_files.INCL_DEFAULT, package)
 agent_out_dir = get_absolute_path(args.agentdir)
 client_out_dir = get_absolute_path(args.clientdir)
+gen_idl = args.gen_idl
+idl_dir = args.idl_dir
+if idl_dir != '':
+    idl_dir = get_absolute_path(args.idl_dir)
+else:
+    idl_dir = agent_out_dir + "/idl"
```
[the remark this answers is not in the document]
Comment (TSC21) [reply msg/tools/generate_microRTPS_bridge.py]: done

File: msg/tools/px_generate_uorb_topic_helper.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (TSC21) [reply msg/tools/px_generate_uorb_topic_helper.py]: done

**Comment by TSC21** [comment]
@bkueng I addressed your comments except the ones related to `uorb_rtps_message_ids.yaml` which I think we can do in a new PR. I also did some beautification of the code according to PEP8. Is that ok with you?

**Comment by TSC21** [comment]
@bkueng I add two more options to the argparser. Can you approve and merge?

**Comment by LorenzMeier** [comment]
Great! Can you please also send a documentation PR for folks who want to use this?

**Comment by TSC21** [comment]
> Great! Can you please also send a documentation PR for folks who want to use this?

Yes. But in this case, what would be the application context? Someone that wants to have a particular implementation of the agent on their system, with some specific agent templates and IDL's? (not including ROS2)

**Comment by TSC21** [comment]
@LorenzMeier had to rebase so to check if CI passes with the new docker container tags on master.

**Comment by TSC21** [comment]
@LorenzMeier the reason: https://github.com/PX4/Firmware/pull/10463#issuecomment-420555650

**Comment by dagar** [comment]
Getting closer. Snapdragon needed an update as well. https://github.com/PX4/Firmware/pull/10468

**Comment by TSC21** [comment]
@dagar OSX is still giving us trouble...

**Comment by dagar** [comment]
That's weird, on that machine (dronecode-macmini) I used this script to update the dependencies yesterday. https://github.com/PX4/Firmware/blob/master/Tools/setup/OSX.sh

Do you have a mac?

**Comment by TSC21** [comment]
> Do you have a mac?

I don't. I think this can be some clash between python stuff installed by pip and by homebrew.

**Comment by TSC21** [comment]
Anyone with a Mac that can have a test to https://github.com/PX4/Firmware/blob/master/Tools/setup/OSX.sh?

**Comment by TSC21** [comment]
@dagar can you please try the changes I put on my latest commit on `dronecode-macmini`? I am not sure about the exports, but they shouldn't be harmful. Also, can you test the output of `which -a pip` and `pip -V`? It may be that it's installing the packages in `/usr/local/lib/python2.7/site-packages`, so that path needs to be added to `PATH`.

**Comment by TSC21** [comment]
> That's weird, on that machine (dronecode-macmini) I used this script to update the dependencies yesterday. https://github.com/PX4/Firmware/blob/master/Tools/setup/OSX.sh

Can you check if pyyaml was correctly installed?

**Comment by LorenzMeier** [comment]
At minimum this is required - but I'm still getting errors, as I had some of the packages already on my system - can fiddle with it later today.

```
diff --git a/Tools/setup/OSX.sh b/Tools/setup/OSX.sh
index d83723cd50..46611673ce 100755
--- a/Tools/setup/OSX.sh
+++ b/Tools/setup/OSX.sh
@@ -8,7 +8,7 @@ brew install px4-dev
 
 # python dependencies
 sudo easy_install pip
-pip install -r ${DIR}/requirements.txt
+sudo pip install -r ${DIR}/requirements.txt
 
 # Optional, but recommended additional simulation tools:
 brew install px4-sim
```

**Comment by TSC21** [comment]
@LorenzMeier any luck in the mean time?

**Comment by TSC21** [comment]
@MaEtUgR are you able to update the windows build in appveyor so CI can pass? (https://ci.appveyor.com/project/MaEtUgR/firmware/build/1458)
Thanks!

**Comment by MaEtUgR** [comment]
Script update (tested to work with this pr): https://github.com/PX4/windows-toolchain/commit/6f752940919211953008cb2b08ba1243fb176a0c
I'm generating the new msi with multiple fixes at the moment. I need to continue scripting and make that fully automated on a cloud server.

**Comment by TSC21** [comment]
> I'm generating the new msi with multiple fixes at the moment. I need to continue scripting and make that fully automated on a cloud server.

Awesome! Thank you! Let me know when I can rebase this so to use the new msi file.

**Comment by MaEtUgR** [comment]
@TSC21 Done: #10516 just cherry-pick my commit to change the toolchain version.

**Comment by TSC21** [comment]
Awesome! Thanks! We are just missing OSX now.

**Comment by MaEtUgR** [comment]
@TSC21 I made #10516 independent such that you don't have to worry and can just rebase. Also it helps if every CI call uses the same file because it gets chached on appveyor.

**Comment by thomasgubler** [comment]
@dagar this is blocking development of the rtsp messaging interface. Would it be possible to give @TSC21  ssh access to update the script?

**Comment by TSC21** [comment]
`dronecode-macmini` CI build fixed. Now missing `gus-macpro`. Already pinged @dogmaphobic about it.
I will then issue a PR with a small update on the `OSX.sh` setup script

**Comment by TSC21** [comment]
@bkueng FYI

**Comment by TSC21** [comment]
@dogmaphobic thanks for updating the machine! CI now passes. This is good to merge.

**Comment by TSC21** [comment]
@bkueng sure thing. Thanks for merging

**Pull request merged**